openedx/core/djangoapps/util/ip.py

```diff
@@ -0,0 +1,186 @@
+            # it's suboptimal. This could actually be a perfectly fine
+            # IP.  Perhaps the server is directly exposed to the
+            # internet (REMOTE_ADDR is a public IP) and the client
+            # sends in a garbage XFF.
+            return current_ip and str(current_ip)
+
+        if next_ip.is_global:
+            # Lacking any knowledge of trusted proxies and other network
+            # configuration, the first public IP we find is the best we
+            # can do.
+            return str(next_ip)
+        else:
+            # If we're still seeing private-space IPs, keep walking;
+            # maybe we'll find a public one! But keep hold of this as
+            # "current"; we may need to fall back to this if the next
+            # one just turns out to be garbage.
+            current_ip = next_ip
+
+    # Ran off the front of the list, so go with what we have. This
+    # would have to be a private IP, or None if the list was in fact
+    # empty to begin with.
+    return current_ip and str(current_ip)
+
+
+def get_client_ip_via_xff(request_meta):
+    """
+    Get the most likely client IP from the X-Forwarded-For chain, conservatively.
+
+    Pick the first publicly routable IP seen while walking from right to left
+    (starting with REMOTE_ADDR), unless the list is exhausted or an invalid IP
+    is found, in which case use the last IP before that.
+    """
+    # The REMOTE_ADDR is implicitly the beginning of the XFF chain
+    # (this is what we would attach if we were a proxy!), so add it on
+    # explicitly.  This is the IP address we can always rely on
+    # having. If we're behind a proxy it may be a private IP, and if
+    # we're serving directly to the internet then it may well be the
+    # client IP.
+    full_chain = get_meta_ips(request_meta, 'X-Forwarded-For') + [request_meta['REMOTE_ADDR']]
+    chosen_ip = conservatively_pick_client_ip(full_chain)
+
+    # In practice this fallback should never happen, since it would require
+    # REMOTE_ADDR (at the end of the chain above) to be malformed.
+    return chosen_ip or request_meta['REMOTE_ADDR']
+
+
+def get_client_ip(request):
+    """
+    Determine the IP address of the HTTP client by walking the X-Forwarded-For
+    header, unless there's a configured override.
+    """
+    # Restore the original REMOTE_ADDR since it's needed for IP determination.
+    # Once XForwardedForMiddleware is no longer overwriting REMOTE_ADDR this
+    # rewriting can be removed.
+    #
+    # This is also the reason all the other functions in this module take a
+    # request.META -- it's easier to pass in an altered META dict than an
+    # altered request, and this keeps the ORIGINAL_REMOTE_ADDR pollution out
+    # of the other code, making it marginally more reusable.
+    request_meta = request.META.copy()
+    if 'ORIGINAL_REMOTE_ADDR' in request_meta:
+        request_meta['REMOTE_ADDR'] = request_meta['ORIGINAL_REMOTE_ADDR']
```

Comment:
1. consider moving this to a new method, and putting the comments into the docstring of the new code.
```
request_meta = get_original_request_meta(request)
```
2. I imagine we only need to make a copy of `META` if we are editing it? The following keeps the conditional code together.
```suggestion
    request_meta = request.META
    if 'ORIGINAL_REMOTE_ADDR' in request_meta:
        request_meta = request.META.copy()
        request_meta['REMOTE_ADDR'] = request_meta['ORIGINAL_REMOTE_ADDR']
```

Reply:
1. I think it's small enough (both in code size and testing surface) that I'd prefer to keep it inlined.
2. I could see doing the copy only if needed, but in practice it will always be needed until `XForwardedForMiddleware` stops doing the overwriting.